07-errors.md

```diff
@@ -0,0 +1,301 @@
+   recorded as the messages exit code.
+3. If a message runs out of gas, the VM will return a receipt with the `ExitCode::SysErrOutOfGas`
+   abnormal exit code.
+
+Changes from pre-FVM Filecoin:
+
+1. Removed `ExitCode::SysErrInvalidMethod` (3). Dispatch is now handled inside actor code.
+2. Removed `ExitCode::SysErrForbidden` (8). Caller verification is a userspace operation, we'll now
+   exit with `ExitCode::ErrForbidden`.
+4. Replaced `ExitCode::SysErrIllegalArgument` with `ExitCode::SysErrAssertionFailed` to more
+   accurately reflect its usage. When this appears on-chain, it means that a message hit some
+   internal Filecoin assert that shouldn't happen.
+
+### Standard "Abnormal Exit" Codes
+
+Standard abnormal exit codes are codes returned by actors to indicate common abnormal exit conditions.
+
+| Code  | Status                 | Description                                         |
+|-------|------------------------|-----------------------------------------------------|
+| 16    | `ErrIllegalArgument`   | invalid message parameters                          |
+| 17    | `ErrNotFound`          | message referenced something that doesn't exist     |
+| 18    | `ErrForbidden`         | operation forbidden                                 |
+| 19    | `ErrInsufficientFunds` | insufficient funds for operation                    |
+| 20    | `ErrIllegalState`      | the actor is in an illegal state                    |
+| 21    | `ErrSerialization`     | actor failed to serialize or deserialize some state |
+| 22    | `ErrInvalidMethod`     | specified method does not exist                     |
+| 23    | `ErrUnspecified`       | the actor failed with an unspecified error          |
+| 24-31 | reserved               |                                                     |
+
+Changes from pre-FVM Filecoin:
+
+1. Added `ExitCode::ErrInvalidMethod` (replacing `ExitCode::SysErrInvalidMethod`).
+2. Added `ExitCode::ErrUnspecified`. This generally means "there's a bug and I have no idea what
+   went wrong". The alternative would be for the actor to _panic_ (leading to an
+   `ExitCode::ErrIllegalActor` syscall error), but it may be useful to distinguish between
+   "crashing" and "something unknown went wrong".
+
+## Syscall Error Numbers
+
+Syscall error numbers (`ErrorNumber`) are returned by syscalls to actors. They indicate that the
+syscall failed (without any side effects). The actor may choose to handle the error and continue, or
+abort (usually with a standard "abnormal exit" code).
+
+A return value of "0" means that the syscall succeeded.
+
+| Number | Name                | Description                                            |
+|--------|---------------------|--------------------------------------------------------|
+| 1      | `IllegalArgument`   | invalid syscall parameters                             |
+| 2      | `IllegalActor`      | actor is not in the correct state to perform operation |
```

Comment:
```suggestion
| 2      | `IllegalActor`      | actor is not in the correct state to perform operation; as opposed to `SysErrIllegalActor` this is a recoverable offence |
```

Reply:
Now: what is the heuristics to decide between one and the other? What is recoverable and what isn't?

Reply:
And shouldn't all such errors just be irrecoverable? What can the actor do here besides rethrowing?

Reply:
> Now: what is the heuristics to decide between one and the other? What is recoverable and what isn't?

At the moment, the only "unrecoverable" error is an abort with a system-reserved exit code (see above for my plans to just rename the exit code.

> And shouldn't all such errors just be irrecoverable? What can the actor do here besides rethrowing?

We can hit this error in two cases:

1. An actor tries to read/write its state after being deleted. There are likely cleanup cases where being able to say "oh, I guess I've already been deleted" would be convenient.
2. An actor attempts to delete itself, with itself as the beneficiary.

In general, my decision process concerning questions like this is:

1. Case 1: We disallow it. We save a tiny bit of gas in an error case that's likely going to fail the entire message.
2. Case 2: We allow it. Common-case, we immediately abort. But if someone _needs_ to handle errors like this, they can.

_However_, this highlights the fact that this error name is... not good. It exists to indicate "this actor was written incorrectly", but I'm not convinced that's what it's going to mean in practice. I'd expect actors to, e.g., call `self_destruct(some_addr)` _without_ checking if `some_addr` is self because:

1. It'll fail either way.
2. it'll cost less gas.

I'm thinking we might want two errors here:

1. `InvalidState`. We'd return this whenever we try to set the state root after deletion, and whenever we attempt to _read_ the state-root inside the constructor (before setting it).
2. `InvalidRecipient`. Distinct from `NotFound` because the recipient _is_ found, just not valid for the current operation. This error is a _bit_ specialized, unfortunately. Maybe `Forbidden` is close enough? We're more likely to have uses for that in the future.

Reply:
Hm. Actually, a general purpose `Forbidden` could also replace `LimitExceeded` (as a matter of fact, we _currently_ use forbidden for that).